core/shared/src/main/scala/cats/parse/Parser.scala

```diff
@@ -72,162 +100,315 @@ sealed abstract class Parser[+A] {
       Left(Parser.Error(offset, Parser.Expectation.unify(NonEmptyList.fromListUnsafe(err.toList))))
   }
 
-  /** If this returns an epsilon failure, change to success with
-    * a value of None, else Some for the parsed value
-    */
+  /**
+   * Convert epsilon failures into None values.
+   *
+   * Normally if a parser fails to consume any input it fails with an
+   * epsilon failure. The `?` method converts these failures into
+   * None values (and wraps other values in `Some(_)`).
+   *
+   * If the underlying parser failed with other errors, this parser
+   * will still fail.
+   */
   def ? : Parser[Option[A]] =
     Parser.oneOf(Parser.map(this)(Some(_)) :: Parser.Impl.optTail)
 
-  /** Don't capture anything, just do the parsing
-    * This can be a significant optimization in that
-    * it removes internal allocations as much as possible.
-    *
-    * see also Functor.as(a) which is .void.map(_ => a)
-    * or a *> b and a <* b which use void to remove allocations
-    * for the side that is discarded.
-    */
+  /**
+   * Parse without capturing values.
+   *
+   * Calling `void` on a parser can be a significant optimization --
+   * it allows the parser to avoid allocating results to return.
+   *
+   * Other methods like `as`, `*>`, and `<*` use `void` internally to
+   * discard allocations, since they will ignore the original parsed
+   * result.
+   */
   def void: Parser[Unit] =
     Parser.void(this)
 
-  /** Discard the value and get the substring that this parser
-    * matches
-    */
+  /**
+   * Return the string matched by this parser.
+   *
+   * When parsing an input string that the underlying parser matches,
+   * this parser will return the matched substring instead of any
+   * value that the underlying parser would have returned. It will
+   * still match exactly the same inputs as the original parser.
+   */
   def string: Parser[String] =
     Parser.string(this)
 
-  /** If this fails, rewind the offset to the starting point
-    * so we can be used with oneOf
-    */
+  /**
+   * If this parser fails to match, rewind the offset to the starting
+   * point before moving on to other parser.
+   *
+   * This method will mostly be used before calling methos such as
+   * `orElse`, `~`, or `flatMap` which involve a subsequent parser
+   * picking up where this one left off.
+   */
   def backtrack: Parser[A] =
     Parser.backtrack(this)
 
+  /**
+   * Sequence another parser after this one, combining both results
+   * into a tuple.
+   *
+   * This combinator returns a product of parsers. If this parser
+   * successfully produces an `A` value, the other parser is run on
+   * the remaining input to try to produce a `B` value.
+   *
+   * If either parser produces an error the result is an error.
+   * Otherwise both extracted values are combined into a tuple.
+   *
+   * Backtracking may be used on the left parser to allow the right
+   * one to pick up after an error.
```

Comment:
This isn't quite right. If A fails B can never run, the only question is will you have an epsilon or arresting failure. If A needs backtracking but B does not, you might do A.backtrack ~ B which is different from (A ~ B).backtrack